### ISSUE TITLE
AWS EKS - add cmd: failed to assign an IP address to container

### ISSUE THREAD

**Issue by djschnei21** [aws/amazon-vpc-cni-k8s#1245]
**What happened**:
Anytime we try to launch a pod with a security group it fails to provision an ENI with the below error.  Also have the cni support bundle if someone would like that via email.

`failed to allocate branch ENI to pod: cannot create new eni entry already exist, older entry : [0xc0002c2310]`

_kubectl_:
`networkPlugin cni failed to set up pod "busybox-test-sg-pod_default" network: add cmd: failed to assign an IP address to container`

_ipamd.log:_
`{"level":"info","ts":"2020-10-01T19:58:56.001Z","caller":"rpc/rpc.pb.go:504","msg":"Received DelNetwork for Sandbox 1097c78a91b063665bed41ce09ca6d4be7cda0375f5546efe4084ce775f040e4"}
{"level":"info","ts":"2020-10-01T19:58:56.004Z","caller":"rpc/rpc.pb.go:504","msg":"Send DelNetworkReply: IPv4Addr , DeviceNumber: 0, err: datastore: unknown pod"}`

**What you expected to happen**:
CNI to attach ENI to pod with Security group
**How to reproduce it (as minimally and precisely as possible)**:
Anytime we try to launch a pod with a security group
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.17
- CNI Version: 1.7.3
- OS (e.g: `cat /etc/os-release`): Amazon Linux 2
- Kernel (e.g. `uname -a`): 4.14.193-149.317.amzn2.x86_64

**Comment by mogren** [comment]
@djschnei21 What instance type is this? Not all instance types support branch ENIs. From [the AWS docs](https://docs.aws.amazon.com/eks/latest/userguide/security-groups-for-pods.html):

> Security groups for pods are supported by most Nitro-based Amazon EC2 instance families, including the m5, c5, r5, p3, m6g, cg6, and r6g instance families. The t3 instance family is not supported. For a complete list of supported instances, see [Amazon EC2 supported instances and branch network interfaces](https://docs.aws.amazon.com/eks/latest/userguide/security-groups-for-pods.html#supported-instance-types). Your nodes must be one of the supported instance types.

If you run `kubectl get nodes -oyaml | grep 'vpc.amazonaws.com/has-trunk-attached'`, what does it say?

**Comment by djschnei21** [comment]
Hey mogren - it's a c5.4xlarge.  here is the output of that command:
`$ kubectl get nodes -oyaml | grep 'vpc.amazonaws.com/has-trunk-attached'
      vpc.amazonaws.com/has-trunk-attached: "true"
      vpc.amazonaws.com/has-trunk-attached: "true"`

**Comment by SaranBalaji90** [comment]
@djschnei21 Does describe pods have any events associated? This could happen if permissions are set appropriately on the cluster role.

**Comment by djschnei21** [comment]
Hey @SaranBalaji90 - the describe pod is showing this error:
`Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "c54935a5c5982d95ef9391fdfff08b86bbf6b8c4b6568b4ee7bb81fce8a0f424" network for pod "busybox-test-sg-pod": networkPlugin cni failed to set up pod "busybox-test-sg-pod_default" network: add cmd: failed to assign an IP address to container`

unfortunately slack is blocked on our work network... but a call would be much appreciated! any chance I can invite you to a teams meeting?

**Comment by djschnei21** [comment]
@SaranBalaji90 we do have an open AWS case if you are able to view those 7447541151 and we can do the call through that maybe?

**Comment by SaranBalaji90** [comment]
Thanks @djschnei21 support ticket link helped. Let me go over the logs attached.

**Comment by hetpats** [comment]
HI @SaranBalaji90  thank you for you help the other day we are still testing various scenarios and  will share the results however we had a had query nd would help us if you can help for below..
1)Do we need to enable DISABLE_TCP_EARLY_DEMUX to true since ENABLE_POD_ENI is set to true?
2)Is there way to disable / block any deployments on cluster that do not have any SG policy assigned ?

**Comment by abhipth** [comment]
Hi @hetpats thanks for reaching out, to answer your questions
1) In order for the TCP connection from Kubelet to pod to succeed the customer must set this DISABLE_TCP_EARLY_DEMUX flag to True. More detailed documentation on this flag and why setting it is required can be found in the [Readme](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/README.md)
2) Currently there is no way to block pods that don't match a SGP. All the pods that will not match any Security Group Policy will not get a dedicated ENI and run with secondary IP Address as a regular pod.

I hope this answers your questions, please feel free to reach out again if you see any issues in your testing or have any further question.

**Comment by PriyaSivaraj1987** [comment]
We are receiving same error for pods after upgrading CNI to 1.17.. Can you please let us know the how to resolve this issue?

**Comment by jayanthvn** [comment]
Hi @PriyaSivaraj1987 

Just to confirm, you are using security group per pods? Can you please confirm the versions of CNI you upgraded?

Thanks.

**Comment by errogaht** [comment]
Hi we experiencing the issue 20 times per day! 
failed to assign an IP address to container

cluster works perfect before. past few days we cant work as normal
re open issue again!

(combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "5b342854fb88ebb0316612ec55c23996277a06b7f4bcd6d26b2276765c8cc128" network for pod "backend-5546c84c9b-tm9mw": networkPlugin cni failed to set up pod "backend-5546c84c9b-tm9mw_staging" network: add cmd: failed to assign an IP address to container

**Comment by errogaht** [comment]
omg i have found the problem, but there is no good logs to fast understand the problem 
![image](https://user-images.githubusercontent.com/6324265/105346370-8a947680-5bf6-11eb-9b66-76f1d58b115f.png)

**Comment by savealive** [comment]
We experience this issue quite often. We do have ENABLE_POD_ENI and DISABLE_TCP_EARLY_DEMUX enabled.
There's enough IPs in subnets
Actually issue has begun to appear when we enabled pod eni.
amazon-k8s-cni:v1.7.5

**Comment by SaranBalaji90** [comment]
@abhipth Can you create a backlog item in https://github.com/aws/amazon-vpc-resource-controller-k8s/issues to emit pod event when ENI cannot be created. This will help debugging the issues.

@savealive, Can you send your EKS cluster ARN to srajakum@amazon.com, I can look into this.

**Comment by savealive** [comment]
@SaranBalaji90 it’s strange and may be a coincidence but upgrading aws-cni to v1.7.8 has solved the issue. I’ll keep an eye and keep you updated whether or not problem really went away.

**Comment by abhipth** [comment]
@SaranBalaji90 Added a backlog item for emitting events in case controller fails to allocate ENI to pod - https://github.com/aws/amazon-vpc-resource-controller-k8s/issues/15

**Comment by alexanderaucobo** [comment]
We also encounter this problem. And start with v1.7.5. 

@savealive is it save to update to that version? Because in the official docs it is 1.7.5

**Comment by savealive** [comment]
Unfortunately upgrade aws-cni to v1.7.8 doesn't solve the issue. 
My observations:
1. Initially there are many pods in state ContainerCreating in a cluster. Many pods are in ContainerCreating for more than 2-3 hours.
```
18m         Warning   FailedCreatePodSandBox   pod/weatherforecastredshift2runforecastiw22-e299a983dcce42399b6077f9cbce3d51   (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "6f1af17f6d90327348a6acb9c3ad5499640121f50acdcdb65a75a9d8d17865ea" network for pod "weatherforecastredshift2runforecastiw22-e299a983dcce42399b6077f9cbce3d51": networkPlugin cni failed to set up pod 
"weatherforecastredshift2runforecastiw22-e299a983dcce42399b6077f9cbce3d51_prod-weatherdata" network: add cmd: 
failed to assign an IP address to container
```
2. Update aws-cni daemonset to v1.7.8 triggers all aws-cni pods to restart.
3. All pods which initially were in ContainerCreating status (no IP from CNI) are successfully got an IP and started up without any actions (i.e. no k delete pod or so)
4. New pods created successfully. Something like `k create deployment nginx --image nginx --replicas 50` works without any issues. 

But next day things are back to "normal" and pods are unable to get IPs from CNI again.

Relevant configuration info:
Typical node labels:
```Name:               ip-10-16-10-214.ec2.internal
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=r5a.xlarge
                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=us-east-1
                    failure-domain.beta.kubernetes.io/zone=us-east-1c
                    jupyter-node-purpose=medium
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=ip-10-16-10-214.ec2.local
                    kubernetes.io/os=linux
                    node.kubernetes.io/instance-type=r5a.xlarge
                    topology.kubernetes.io/region=us-east-1
                    topology.kubernetes.io/zone=us-east-1c
                    vpc.amazonaws.com/has-trunk-attached=true
Annotations:        node.alpha.kubernetes.io/ttl: 0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
```
aws-cni DS config:
```
      containers:
      - env:
        - name: ADDITIONAL_ENI_TAGS
          value: '{}'
        - name: AWS_VPC_CNI_NODE_PORT_SUPPORT
          value: "true"
        - name: AWS_VPC_ENI_MTU
          value: "9001"
        - name: AWS_VPC_K8S_CNI_CONFIGURE_RPFILTER
          value: "false"
        - name: AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG
          value: "false"
        - name: AWS_VPC_K8S_CNI_EXTERNALSNAT
          value: "false"
        - name: AWS_VPC_K8S_CNI_LOGLEVEL
          value: DEBUG
        - name: AWS_VPC_K8S_CNI_LOG_FILE
          value: /host/var/log/aws-routed-eni/ipamd.log
        - name: AWS_VPC_K8S_CNI_RANDOMIZESNAT
          value: prng
        - name: AWS_VPC_K8S_CNI_VETHPREFIX
          value: eni
        - name: AWS_VPC_K8S_PLUGIN_LOG_FILE
          value: /var/log/aws-routed-eni/plugin.log
        - name: AWS_VPC_K8S_PLUGIN_LOG_LEVEL
          value: DEBUG
        - name: DISABLE_INTROSPECTION
          value: "false"
        - name: DISABLE_METRICS
          value: "false"
        - name: ENABLE_POD_ENI
          value: "true"
        - name: MY_NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        - name: WARM_ENI_TARGET
          value: "1"
        image: 602401143452.dkr.ecr.us-east-1.amazonaws.com/amazon-k8s-cni:v1.7.8
```
Failed pods DO NOT use SecurityGroups for pods. We use SGs for pods just for few workloads and there's no correlation between failed pods and enabled SGs for pods.

We have 20+ EKS clusters in different accounts with similar configs and looks like all of them are affected.
Most of our clusters are on EKS 1.18, aws-cni v1.7.5.
Issue appears only with pods assigned to specific nodes so we considered that nodes as "failed"
As a temporary workaround we used to drain "failed" nodes and replace them with new ones. In most cases "new" nodes were ok and pods on them got IP-address successfully.

**Comment by SaranBalaji90** [comment]
@savealive we are looking into this to investigate further. If this problem occurs on any of your node again, can you collect the logs using https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html#troubleshoot-cni and send it to us. It will be super useful for debugging this further.

**Comment by SaranBalaji90** [comment]
Based on the logs I was able to narrow down to following things
 
1. To use per pod security group, EKS cluster should use aws-cni version >= v1.7.7 (https://github.com/aws/amazon-vpc-cni-k8s/pull/1318) as we fixed bug in the deletion path which was not freeing up the IPs. I will make sure our docs are updated to reflect this.
2. This page shows list of supported instance types - https://docs.aws.amazon.com/eks/latest/userguide/security-groups-for-pods.html#supported-instance-types. To enable per pod security group, EKS Cluster should use one of the supported instance types for all the nodes. Created request to support on clusters which has some unsupported worker nodes as well - https://github.com/aws/amazon-vpc-cni-k8s/issues/1363.

But please do send me logs for nodes using 1.7.8 and having issues.

**Comment by SaranBalaji90** [comment]
Just to give an update here - Myself and @jayanthvn reviewed the log files and we found the maxPod is set incorrectly on cluster with custom networking (ENIConfig) enabled. More details about this feature is documented here - https://docs.aws.amazon.com/eks/latest/userguide/cni-custom-network.html. Let us know if problem still exist we can dive deep further :)

**Comment by danghung-dev** [comment]
I have the same problem.
`
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "76489f2063cb456952195756954f82b49a3f6ff73708cdc39d8b390ab593d50e" network for pod "swm-be-dev-7bcdd4c64f-69nv7": networkPlugin cni failed to set up pod "swm-be-dev-7bcdd4c64f-69nv7_swm" network: add cmd: failed to assign an IP address to container
`

Our environment is
ẸKS 1.19
AWS-CNI v1.7.5

Now i cannot deploy new pod because of this problem.

**Comment by adhityan** [comment]
I have updated my cni to 1.7.9 and still seeing this error -

Environment is
ẸKS 1.19
AWS-CNI v1.7.9
Uses Fargate + NodeGroups

Error:
`Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "89a7c4d0455147ea9eeece8b9bd12059835b40cec28be6b1b4988ce69fc312a8" network for pod "insta-service-6b959c8fdd-cw5pm": networkPlugin cni failed to set up pod "insta-service-6b959c8fdd-cw5pm_insta-service" network: add cmd: failed to assign an IP address to container`

**Comment by rohith-gittest** [comment]
Issue Summary:
EKS worker nodes has limitations  for running number of pod based on the instance type. In our case we use instance type-c5.2xlarge which allows 58 pods per instance. If the pods are scheduled more than that number on the same instance, then those pods are getting stuck in ContainerCreating status.
Describing the pods gives us following errors:  " Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "8be3026cc6bb2b311570f118dfef8ab93ae491e6d6fc20e12a46e9b814cff716" network for pod "pod-name": networkPlugin cni failed to set up pod "pod-name" network: add cmd: failed to assign an IP address to container "

As part of solution we tried adding --use-max-pods false --kubelet-extra-args '--max-pods=110' into our EKS worker node bootstrap script. Still the issue remains same.
Please guide us the solutions if there are any to remove pods restriction based on instance type and run max number of pods in a worker node.

Note: 
Based on the AWS premium support they suggest us to consider greater instance types to achieve our use-case(which is to run max number of pods in a worker node). 
This solution provided by AWS is absolutely weird in many ways as listed below:

1) If we run maximum number of pods in a EKS worker node, Still we can see worker nodes are being underutilized in terms of CPU and memory. In this case using higher instance type will cost us more.
Even if i run max number of pods in a worker node consuming the total % of CPU will not cross double digits. In this use case again having greater instance type is not the cost optimization strategy.

2) If the pods are stuck in ContainerCreating state, the same pod can't move to running state unless it's being killed and restarted.
So our requirement is kube-scheduler should know if the node have the sufficient IPs, Then only it should schedule the pod  on that particular node. Indeed kube-scheduler should know if IPs are exhausted in the worker node before placing the pods.

3) If we use aws vpc custom cni then worker nodes cant make use of all the NIC available for that instance type.
For Ex: We use c5.2xlarge which supports max 4 NIC. Since we use a different subnets each for nodes and pods. 1 NIC will be assigned to our node which will have only 1 IP and rest of the 14 IPS will be wasted.
other 3 NIC*15=45 pods only we can run instead of 58 pods. Please give us a justification on this.

4) If pods stuck in Container creating status any auto scaling option available to bring new nodes	as like cluster autoscaler will monito CPU and memory.

**Comment by bfaverdin-yseop** [comment]
Hello,

we have the same issue here:

EKS 1.18
AWS-CNI v1.7.8
Managed nodegroup (t3.2xlarge) with cluster autoscaler 1.18.3
No custom networking

We have the problem on (at least) two clusters, with 3 nodes affected (out of 11) on the production cluster and 1 (out of 2) on the preprod cluster.

Example of kubectl event on a failed pod:

```
42m         Warning   FailedCreatePodSandBox   pod/manager-698857df9f-f2wq7                             Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "a28c3fbd9047867d8bf241b0dc3f394d5b25347b64740fd54fac7dd8071f0bba" network for pod "manager-698857df9f-f2wq7": networkPlugin cni failed to set up pod "manager-698857df9f-f2wq7_afa-1qupv9aak-e5meag-44s0wf73" network: add cmd: failed to assign an IP address to container
```

`ipamd.log`
[ipamd-2021-03-13T15-19-35.224.log.gz](https://github.com/aws/amazon-vpc-cni-k8s/files/6189845/ipamd-2021-03-13T15-19-35.224.log.gz)
 reports that kind of error (see full log file for more).

```
{"level":"debug","ts":"2021-03-06T01:56:17.481Z","caller":"rpc/rpc.pb.go:504","msg":"DelNetworkRequest: ClientVersion:\"v1.7.8\" K8S_POD_NAME:\"acp-hook9a48b66-postsync--62135596800-bdpmz\" K8S_POD_NAMESPACE:\"afa-2snsozaaw-80reaq-p5kr65y2\" K8S_POD_INFRA_CONTAINER_ID:\"a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6\" Reason:\"PodDeleted\" ContainerID:\"a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6\" IfName:\"eth0\" NetworkName:\"aws-cni\" "}
{"level":"debug","ts":"2021-03-06T01:56:17.481Z","caller":"ipamd/rpc_handler.go:196","msg":"UnassignPodIPv4Address: IP address pool stats: total:56, assigned 49, sandbox aws-cni/a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6/eth0"}
{"level":"info","ts":"2021-03-06T01:56:17.481Z","caller":"datastore/data_store.go:786","msg":"UnAssignPodIPv4Address: Unassign IP 192.168.70.57 from sandbox aws-cni/a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6/eth0"}
{"level":"info","ts":"2021-03-06T01:56:17.481Z","caller":"ipamd/rpc_handler.go:196","msg":"UnassignPodIPv4Address: sandbox aws-cni/a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6/eth0's ipAddr 192.168.70.57, DeviceNumber 1"}
{"level":"info","ts":"2021-03-06T01:56:17.481Z","caller":"rpc/rpc.pb.go:504","msg":"Send DelNetworkReply: IPv4Addr 192.168.70.57, DeviceNumber: 1, err: <nil>"}
{"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:857","msg":"IP pool stats: total = 56, used = 48, c.maxIPsPerENI = 14"}
{"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:482","msg":"IP pool is too low: available (8) < ENI target (1) * addrsPerENI (14)"}
{"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:483","msg":"Starting to increase IP pool size"}
{"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:483","msg":"Skipping ENI allocation as the max ENI limit of 4 is already reached (accounting for 0 unmanaged ENIs and 0 trunk

```

FYI the full log file provided corresponds to the logs of a new fresh node created at EKS 1.18 cluster upgrade.

Tested: aws-node restart, no change. Same for the upgrade to v1.7.10 of AWS-CNI, issue is still present.

I can provide the full dump of `/opt/cni/bin/aws-cni-support.sh` on request.

Thanks for your help.

**Comment by SaranBalaji90** [comment]
@bfaverdin-yseop can you please send us the full dump of support script to fawadkh@amazon.com/varavaj@amazon.com.

**Comment by bfaverdin-yseop** [comment]
@SaranBalaji90 Done. Thanks.

**Comment by jayanthvn** [comment]
Hi @bfaverdin-yseop 

I checked the ipam.json file and see there are 48 pods. So I assume you want to know why the remaining 8 IPs are not utilized?

Thank you!

**Comment by bfaverdin-yseop** [comment]
Hi @javanthvn Our main concern is that sometimes pods are scheduled on nodes where no IP is available (or reported as it) and get stuck with `ContainerCreating` status. Pods are not automatically rescheduled on another node in that case and get stuck until they're manually killed and automatically redeployed on another node (if there's enough room to do that).

I checked the node status on our production cluster this morning and allocated IPs has increased from 48 to 55 on the node I gave you the logs, even if errors are still reported:

```
{"level":"debug","ts":"2021-03-24T10:45:08.108Z","caller":"ipamd/ipamd.go:476","msg":"Successfully Reconciled ENI/IP pool"}
{"level":"debug","ts":"2021-03-24T10:45:08.108Z","caller":"ipamd/ipamd.go:476","msg":"IP Address Pool stats: total: 56, assigned: 55"}
{"level":"debug","ts":"2021-03-24T10:45:10.608Z","caller":"ipamd/ipamd.go:857","msg":"IP pool stats: total = 56, used = 55, c.maxIPsPerENI = 14"}
{"level":"debug","ts":"2021-03-24T10:45:10.608Z","caller":"ipamd/ipamd.go:482","msg":"IP pool is too low: available (1) < ENI target (1) * addrsPerENI (14)"}
{"level":"debug","ts":"2021-03-24T10:45:10.608Z","caller":"ipamd/ipamd.go:483","msg":"Starting to increase IP pool size"}
{"level":"debug","ts":"2021-03-24T10:45:10.608Z","caller":"ipamd/ipamd.go:483","msg":"Skipping ENI allocation as the max ENI limit of 4 is already reached (accounting for 0 unmanaged ENIs and 0 trunk ENIs)"}
```

The two other nodes reporting errors have also 55 IPs allocated (no change), as well as other nodes with 55 IPs allocated but not reporting any error.

I've done some extra tests. As new pods have been created since this morning our cluster is now nearly full (only 2 IPs available). I restarted two deployments (14 new pods created) to test if some pods got stuck at creation. Pods remained with `Pending` state until the cluster autoscaler created a new node to schedule the pods, which is the expected behavior, but no pod got stuck with `ContainerCreating` status. Unfortunately it seems I'm not able to reproduce the issue, as things behave as expected, even if no change has been done to the cluster (tests on aws-node were done on the preprod cluster).

**Comment by jayanthvn** [comment]
Hi @rohith-gittest 

Sorry for the delayed response, yes today with 1.X max-pods depends on the number of IPs/ENI limits. But we are working on increasing the pod density - https://github.com/aws/containers-roadmap/issues/138 and this should help you pack more pods better. Reg point 3, we won't use Primary ENI since the ENI will be in node subnet and other ENIs will be in another subnet passed by ENIConfig which will be used for pods. Reg point 2, max pods are computed based on number of IPs so kube-scheduler should take that into account while assigning a pod to the node. Hope this helps.

**Comment by rohith-gittest** [comment]
Hi @jayanthvn 
Reg point 2, Could you please tell us why kube-scheduler doesn't schedule the pods by considering the number of IPs available in the nodes. Give us justification or is it will be part of improvements in the upcoming release.

**Comment by jayanthvn** [comment]
Hi @rohith-mr-rao 

Kube-scheduler that is running in EKS is vanilla.  Factors that need taken into account for scheduling decisions include individual and collective resource requirements, hardware / software / policy constraints, affinity and anti-affinity specifications, data locality, inter-workload interference, and so on but does not known AWS EKS specific implementations like number of IP address available on the node.

**Comment by sidewinder12s** [comment]
Hi @jayanthvn, I'm seeing similar errors to the original poster however I do not believe we use pod security groups extensively or at all (though we have custom networking enabled and I believe we're setup to support it, I just don't think we have any workloads using it). From the read of the logs, it's like the node starts and then takes 10 minutes before any IPs are available for use from ipamd and during those 10 minutes pods attempt to get scheduled and obviously fail.

Versions/settings:
Kube: 1.18
CNI Plugin: 1.7.5
Instance: r4.16xlarge (The logs are from this instance type, though we also have r5, p2 probably a few other generations in this cluster)
Non-Default CNI Settings (through the helm chart):

```yaml
            - name: ADDITIONAL_ENI_TAGS
              value: '{}'
            - name: AWS_VPC_CNI_NODE_PORT_SUPPORT
              value: 'true'
            - name: AWS_VPC_ENI_MTU
              value: '9001'
            - name: AWS_VPC_K8S_CNI_CONFIGURE_RPFILTER
              value: 'false'
            - name: AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG
              value: 'true'
            - name: AWS_VPC_K8S_CNI_EXTERNALSNAT
              value: 'false'
            - name: AWS_VPC_K8S_CNI_LOGLEVEL
              value: INFO
            - name: AWS_VPC_K8S_CNI_LOG_FILE
              value: stdout
            - name: AWS_VPC_K8S_CNI_RANDOMIZESNAT
              value: prng
            - name: AWS_VPC_K8S_CNI_VETHPREFIX
              value: eni
            - name: AWS_VPC_K8S_PLUGIN_LOG_FILE
              value: stderr
            - name: AWS_VPC_K8S_PLUGIN_LOG_LEVEL
              value: INFO
            - name: DISABLE_INTROSPECTION
              value: 'false'
            - name: DISABLE_METRICS
              value: 'false'
            - name: ENABLE_POD_ENI
              value: 'false'
            - name: ENI_CONFIG_LABEL_DEF
              value: failure-domain.beta.kubernetes.io/zone
            - name: MINIMUM_IP_TARGET
              value: '10'
            - name: WARM_IP_TARGET
              value: '5'
```

I've opened a case with Support about this as we only started noticing this after we messed with the CNI MIN_IP and WARM_IP settings to better utilize our IP space as we have an extremely elastic cluster, though fairly low pod density per node (mainly constrained by GPU availability). We had first actually run out of IPs after scaling out a single AZ to our max ASG size. 

I've recently updated that case with at least I believe full INFO level logs from a node we found effected (running the debug script would be difficult as most of our problem nodes will end up getting deleted before we notice the issue, usually within an hour or 2 of launch), if you could take a look if there is anything obviously wrong? Case ID: 8201085041

**Comment by jayanthvn** [comment]
Hi @sidewinder12s 

Sure I will check the logs attached to the case and reach out to you.

Thanks!

**Comment by jayanthvn** [comment]
> Hello,
> 
> we have the same issue here:
> 
> EKS 1.18
> AWS-CNI v1.7.8
> Managed nodegroup (t3.2xlarge) with cluster autoscaler 1.18.3
> No custom networking
> 
> We have the problem on (at least) two clusters, with 3 nodes affected (out of 11) on the production cluster and 1 (out of 2) on the preprod cluster.
> 
> Example of kubectl event on a failed pod:
> 
> ```
> 42m         Warning   FailedCreatePodSandBox   pod/manager-698857df9f-f2wq7                             Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "a28c3fbd9047867d8bf241b0dc3f394d5b25347b64740fd54fac7dd8071f0bba" network for pod "manager-698857df9f-f2wq7": networkPlugin cni failed to set up pod "manager-698857df9f-f2wq7_afa-1qupv9aak-e5meag-44s0wf73" network: add cmd: failed to assign an IP address to container
> ```
> 
> `ipamd.log` [ipamd-2021-03-13T15-19-35.224.log.gz](https://github.com/aws/amazon-vpc-cni-k8s/files/6189845/ipamd-2021-03-13T15-19-35.224.log.gz) reports that kind of error (see full log file for more).
> 
> ```
> {"level":"debug","ts":"2021-03-06T01:56:17.481Z","caller":"rpc/rpc.pb.go:504","msg":"DelNetworkRequest: ClientVersion:\"v1.7.8\" K8S_POD_NAME:\"acp-hook9a48b66-postsync--62135596800-bdpmz\" K8S_POD_NAMESPACE:\"afa-2snsozaaw-80reaq-p5kr65y2\" K8S_POD_INFRA_CONTAINER_ID:\"a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6\" Reason:\"PodDeleted\" ContainerID:\"a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6\" IfName:\"eth0\" NetworkName:\"aws-cni\" "}
> {"level":"debug","ts":"2021-03-06T01:56:17.481Z","caller":"ipamd/rpc_handler.go:196","msg":"UnassignPodIPv4Address: IP address pool stats: total:56, assigned 49, sandbox aws-cni/a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6/eth0"}
> {"level":"info","ts":"2021-03-06T01:56:17.481Z","caller":"datastore/data_store.go:786","msg":"UnAssignPodIPv4Address: Unassign IP 192.168.70.57 from sandbox aws-cni/a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6/eth0"}
> {"level":"info","ts":"2021-03-06T01:56:17.481Z","caller":"ipamd/rpc_handler.go:196","msg":"UnassignPodIPv4Address: sandbox aws-cni/a1de58c9548a1007cddb5882d2a48405e66837568ff040518afcf24ea7181ba6/eth0's ipAddr 192.168.70.57, DeviceNumber 1"}
> {"level":"info","ts":"2021-03-06T01:56:17.481Z","caller":"rpc/rpc.pb.go:504","msg":"Send DelNetworkReply: IPv4Addr 192.168.70.57, DeviceNumber: 1, err: <nil>"}
> {"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:857","msg":"IP pool stats: total = 56, used = 48, c.maxIPsPerENI = 14"}
> {"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:482","msg":"IP pool is too low: available (8) < ENI target (1) * addrsPerENI (14)"}
> {"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:483","msg":"Starting to increase IP pool size"}
> {"level":"debug","ts":"2021-03-06T01:56:22.018Z","caller":"ipamd/ipamd.go:483","msg":"Skipping ENI allocation as the max ENI limit of 4 is already reached (accounting for 0 unmanaged ENIs and 0 trunk
> ```
> 
> FYI the full log file provided corresponds to the logs of a new fresh node created at EKS 1.18 cluster upgrade.
> 
> Tested: aws-node restart, no change. Same for the upgrade to v1.7.10 of AWS-CNI, issue is still present.
> 
> I can provide the full dump of `/opt/cni/bin/aws-cni-support.sh` on request.
> 
> @SaranBalaji90 Thanks for your help.

Hi,

I did check one of the logs shared in the support ticket, `ipamd.json` shows up 56 pods and even IPAMD is reporting all available IPs are consumed - 

```
{"level":"debug","ts":"2021-04-12T02:45:13.054Z","caller":"ipamd/ipamd.go:476","msg":"Reconciling ENI/IP pool info because time since last 1m0.013309204s <= 1m0s"}
{"level":"debug","ts":"2021-04-12T02:45:13.082Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-0ad1105366bef478a IP pool"}
{"level":"debug","ts":"2021-04-12T02:45:13.082Z","caller":"ipamd/ipamd.go:1028","msg":"Reconcile and skip primary IP 192.168.85.1 on ENI eni-0ad1105366bef478a"}
{"level":"debug","ts":"2021-04-12T02:45:13.082Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-07ba201a3aa81b1d0 IP pool"}
{"level":"debug","ts":"2021-04-12T02:45:13.082Z","caller":"ipamd/ipamd.go:1028","msg":"Reconcile and skip primary IP 192.168.64.232 on ENI eni-07ba201a3aa81b1d0"}
{"level":"debug","ts":"2021-04-12T02:45:13.082Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-05a7d42924c370891 IP pool"}
{"level":"debug","ts":"2021-04-12T02:45:13.083Z","caller":"ipamd/ipamd.go:1028","msg":"Reconcile and skip primary IP 192.168.73.59 on ENI eni-05a7d42924c370891"}
{"level":"debug","ts":"2021-04-12T02:45:13.083Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-0f24156bb32d6ac9f IP pool"}
{"level":"debug","ts":"2021-04-12T02:45:13.083Z","caller":"ipamd/ipamd.go:1028","msg":"Reconcile and skip primary IP 192.168.69.118 on ENI eni-0f24156bb32d6ac9f"}
{"level":"debug","ts":"2021-04-12T02:45:13.083Z","caller":"ipamd/ipamd.go:476","msg":"Successfully Reconciled ENI/IP pool"}
{"level":"debug","ts":"2021-04-12T02:45:13.083Z","caller":"ipamd/ipamd.go:476","msg":"IP Address Pool stats: total: 56, assigned: 56"}
```

**Comment by Nurlan199206** [comment]
i have same issue...

```
kubectl describe daemonset aws-node -n kube-system | grep Image | cut -d "/" -f 2
amazon-k8s-cni-init:v1.7.10
amazon-k8s-cni:v1.7.5-eksbuild.1
```


`Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "58e46f41072c9382edc2d9a2c58eb49acb5b388fb31b99a075ffdd0aa2d796d0" network for pod "frontend-657fddbfbb-gjxhw": networkPlugin cni failed to set up pod "frontend-657fddbfbb-gjxhw_a-plus" network: add cmd: failed to assign an IP address to container`

**Comment by jayanthvn** [comment]
> amazon-k8s-cni-init:v1.7.10
> amazon-k8s-cni:v1.7.5-eksbuild.1

Hi @Nurlan199206 

Can you please open a new issue and share ipamd.log?

**Comment by Nurlan199206** [comment]
@jayanthvn 

i changed this settings in aws-node DaemonSet pods from `1` to `2` and pods have started, maybe it's helped?

```
            - name: WARM_ENI_TARGET
              value: '2'
```

**Comment by Nurlan199206** [comment]
@jayanthvn 

https://github.com/aws/amazon-vpc-cni-k8s/issues/1480

**Comment by dgoscn** [comment]
Does anyone have a solution or a clean solution besides than the weird aws response on @rohith-mr-rao?

**Comment by connatix-cradulescu** [comment]
So the issue occurs in the situation you are using the CNI AWS plugin.  From the start I can tell you is not a technical issue but more a planning issue. You did not take into consideration that the growth of your cluster will drain out the subnet(s) of the cluster.  


How this error appears:

The node that is the pod scheduled for is on the cluster for a certain period of time, let's say  1 day.  In the meantime, your cluster drains out all your IPS from subnet(s). A new pod is scheduled on the node into the discussion and the error will pop up, because you don't have any more IPs on the subnet(s).  The pod is asking for IPs from the  VPC/subnet(s). 

I had the same issue when I did overscale my cluster over 1k nodes. 

Solution:
Create a second subnet, label the subnet with the right labels for EKS ( don't try to modify the master will ask for recreation and you don't want to do that). All you need is to put the Kubernetes labels on the subnets. Then you need to create a lunch configuration to lunch instances from the newly added subnet.  Here you can play with some logic, to spread the load over your subnets. Anyhow, the issue is not the CNI but the lack of IP on the AWS subnet that is used by the Kubernetes cluster.

**Comment by hhannani** [comment]
I faced similar issue, and after searching online, I found this fix as my coredns configmap has upstream line in it. AWS: https://docs.aws.amazon.com/eks/latest/userguide/managing-coredns.html#updating-coredns-eks-add-on

**Comment by shahid-devops** [comment]
Hey Guys,
I am also facing the same issue, I saw the subnets and it has more than 45 available IP's but still i am facing the issue 

Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "d31a2f64c54c259266f4e3eca18b5cc67ec104b5abd612e422c2acfcfecc86bc" network for pod "test-pod--8sq97": networkPlugin cni failed to set up pod "test-pod--8sq97_default" network: add cmd: failed to assign an IP address to container

**Comment by dimitarshenkov** [comment]
Same here CNI plugin version amazon-k8s-cni:v1.10.1-eksbuild.1

**Comment by leviinz** [comment]
Not sure if this will help anyone, but might as well comment here as I spent hours trying to debug this annoying error, regardless of knowing that my subnets *do* have enough addresses.

You need to use an instance type that has IsTrunkingCompatible: true in it in this source file: https://github.com/aws/amazon-vpc-resource-controller-k8s/blob/master/pkg/aws/vpc/limits.go

Once you're using a compatible instance type you'll need to re-create the node group and then restart any failed deployments. If for example coredns failed to start (shows as "degraded" on the EKS dashboard), you can restart it with `kubectl rollout restart -n kube-system deployment/coredns`

**Comment by cslovell** [comment]
Any solution yet so far that doesn't involve increasing node sizes? 

This is a debilitating problem for us. We previously used GKE for our clusters and moved to AWS recently. The very fact that this is not treated as a serious issue with EKS is revealing about how little interest they have in making EKS attractive to AWS users. We will need to find a solution to this, but it is inconceivable that users should start paying for larger machines that they don't really need.

**Comment by jayanthvn** [comment]
@cslovell - There is an ENI and IP limit per instance type and that dictates the the max IPs per node limits -  https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#AvailableIpPerENI. You can enable prefix delegation (https://aws.amazon.com/blogs/containers/amazon-vpc-cni-increases-pods-per-node-limits/) from CNI 1.9.0+ to achieve higher pod density per node - https://aws.amazon.com/blogs/containers/amazon-vpc-cni-increases-pods-per-node-limits/.

**Comment by JhonnyTheRipper** [comment]
I know i'm necroposting, but none of the solutions work
ENABLE_PREFIX_DELEGATION=true
WARM_PREFIX_TARGET=1
IP reservation in subnets
I even moved the entire cluster to m7g as per @levi-nz suggestion

All the subnets have at least 200 available addresses
And yet, I still have the same error
Any suggestions?

**Comment by OneideLuizSchneider** [comment]
@JhonnyTheRipper 
It worked on our EKS clusters (version 1.26 and 1.29)
Command:
`kubectl set env daemonset aws-node -n kube-system ENABLE_PREFIX_DELEGATION=true`

Just tested it again last week.

**Comment by JhonnyTheRipper** [comment]
@OneideLuizSchneider 
Yes, it's the first thing I did.
In fact, aws can't even start ebs-csi and coredns pods (as in on the fresh node) due to the error.

**Comment by JhonnyTheRipper** [comment]
Ok, the issue is I don't have /28 ranges in the subnets.
Trunkated works, after setting the env `kubectl set env daemonset aws-node -n kube-system ENABLE_POD_ENI=true`, but with m7g you are only getting 8 addresses.
The only solution is to create new subnets.

**Comment by kollzey** [comment]
After setting ENABLE_PREFIX_DELEGATION=true, dont forget to also set max pods for kubelet. If you're using launch templates or managed nodes, append these args to startup scripts

/etc/eks/bootstrap.sh my-cluster \
  --use-max-pods false \
  --kubelet-extra-args '--max-pods=110'

**Comment by JnMik** [comment]
Had the issue, to me the problem was that I was using an r6g.medium, which allows only 8 pods in vpc-cni. I had a bunch of daemonset already taking IPs spot on the nodes, making difficult my contains to deploy.

This document has been very useful to me https://github.com/aws/amazon-vpc-cni-k8s/blob/master/misc/eni-max-pods.txt#L699

**Comment by JhonnyTheRipper** [comment]
> Had the issue, to me the problem was that I was using an r6g.medium, which allows only 8 pods in vpc-cni. I had a bunch of daemonset already taking IPs spot on the nodes, making difficult my contains to deploy.
> 
> This document has been very useful to me https://github.com/aws/amazon-vpc-cni-k8s/blob/master/misc/eni-max-pods.txt#L699

You can enable prefix delegation, r6g.medium is supported. That would yield you up to 98 pods

**Comment by OneideLuizSchneider** [comment]
> Ok, the issue is I don't have /28 ranges in the subnets. Trunkated works, after setting the env `kubectl set env daemonset aws-node -n kube-system ENABLE_POD_ENI=true`, but with m7g you are only getting 8 addresses. The only solution is to create new subnets.

well, we are using the m7g as well, and working just fine. 
Btw, EKS 1.31
